ast/src/parsed/mod.rs

```diff
@@ -159,129 +154,21 @@ pub struct PolynomialName<T> {
-    index: Option<Box<Expression<T>>>,
-    /// The underlying polynomial
-    pol: PolynomialReference,
-}
-
-impl<T> IndexedPolynomialReference<T> {
-    /// Return a namespaced polynomial based on this polynomial and an optional namespace
-    pub fn with_namespace(self, namespace: Option<String>) -> NamespacedPolynomialReference<T> {
-        NamespacedPolynomialReference {
-            pol: self,
-            namespace,
-        }
-    }
-
-    /// Returns a mutable reference to the name of the declared polynomial or array of polynomials
-    pub fn name_mut(&mut self) -> &mut String {
-        self.pol.name_mut()
-    }
-
-    /// Returns the optional index of this polynomial in its declaration array
-    pub fn index(&self) -> &Option<Box<Expression<T>>> {
-        &self.index
-    }
-
-    /// Returns the name of the declared polynomial or array of polynomials
-    pub fn name(&self) -> &str {
-        self.pol.name()
-    }
-
-    /// Return a namespaced polynomial based on this polynomial and a namespace
-    pub fn namespaced(self, namespace: String) -> NamespacedPolynomialReference<T> {
-        self.with_namespace(Some(namespace))
-    }
-
-    /// Return a namespaced polynomial based on this polynomial and no namespace, defaulting to the closest enclosing namespace, if any
-    pub fn local(self) -> NamespacedPolynomialReference<T> {
-        self.with_namespace(None)
-    }
-}
-
-#[derive(Debug, PartialEq, Eq, Default, Clone, PartialOrd, Ord)]
-/// A polynomial or array of polynomials
-pub struct PolynomialReference {
-    /// The name of this polynomial or array of polynomials
-    name: String,
-}
-
-impl PolynomialReference {
-    /// Returns an indexed polynomial using this polynomial and an optional index
-    pub fn with_index<T>(self, index: Option<Expression<T>>) -> IndexedPolynomialReference<T> {
-        IndexedPolynomialReference {
-            pol: self,
-            index: index.map(Box::new),
-        }
-    }
-
-    /// Returns the name of this polynomial or array of polynomials
-    pub fn name(&self) -> &str {
-        &self.name
-    }
-
-    /// Returns a mutable reference to the name of this polynomial or array of polynomials
-    pub fn name_mut(&mut self) -> &mut String {
-        &mut self.name
-    }
-
-    /// Returns a new polynomial or array of polynomials given a name
-    pub fn new<S: Into<String>>(name: S) -> Self {
-        Self { name: name.into() }
-    }
-
-    /// Returns an indexed polynomial using this polynomial and an index. Used for polynomial array members.
-    pub fn indexed<T>(self, index: Expression<T>) -> IndexedPolynomialReference<T> {
-        self.with_index(Some(index))
-    }
-
-    /// Returns an indexed polynomial using this polynomial and an index. Used for polynomials which are not declared in arrays.
-    pub fn single<T>(self) -> IndexedPolynomialReference<T> {
-        self.with_index(None)
-    }
```

Comment:
Why is none of this needed anymore?

Reply:
This changes `[]` to a proper operator. It allow things like `f(x)[2]` or `(f + 2)[3]`. Previously, index access was tied to a symbol, and this is the reflection of that in the rust type system.